include/mqtt/endpoint.hpp

```diff
@@ -1466,17 +1470,14 @@ class endpoint : public std::enable_shared_from_this<endpoint<Mutex, LockGuard,
      *        Properties<BR>
      *        See https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901145<BR>
      *        3.6.2.2 PUBREL Properties
-     * @param life_keeper
-     *        An object that stays alive (but is moved with force_move()) until the async operation is finished.
      * @param
      */
     void pubrel(
         packet_id_t packet_id,
         v5::pubrel_reason_code reason_code = v5::pubrel_reason_code::success,
-        v5::properties props = {},
-        any life_keeper = any()
+        v5::properties props = {}
```

Comment:
You removed life_keeper here. I agree with the removal. But I'd like to clarify the following point.

* `props` must manage their lifetime by themselves.

`v5::properties` has `v5::property_variant`. The actual type of `v5::property_variant` could be a string property such as `user_property`. The string property has `buffer`. The `buffer` has `const_shared_ptr_array`. It can be null.

If user create intentionally the reference type of `buffer` and pass it to the string property, then pass the string property to pubrel() as the part of the props, then the (removed) parameter life_keeper can be used to keep the lifetime of the `buffer`. But I think that it is unexpected usage. It is a kind of abuse. (But not documented well)

When properties are received by `mqtt_cpp`, then the properties might be share their lifetime. Each `buffer`'s `const_shared_ptr_array` are the same. In this case, there is no problem to pass to the properties as the parameter of pubrel().

Reply:
I'm not sure it's a good idea to remove lifetime_keeper from the pubrel functions. 

I want to support transmitting / receiving properties without memory allocations, so I'd rather keep the life_keeper.

Thank you for pointing out that change. I will revert it.

Reply:
Do you mean we need to keep life_keeper parameter for props?
If you do, storing message is not only pubrel but also publish.

Now we have three publish overloads:

```cpp
    void publish(packet_id_t packet_id, std::string topic_name, std::string contents, publish_options pubopts = {}, v5::properties props = {});
    void publish(packet_id_t packet_id, buffer topic_name, buffer contents, publish_options pubopts = {}, v5::properties props = {});
    void publish(packet_id_t packet_id, as::const_buffer topic_name, as::const_buffer contents, MQTT_NS::any life_keeper, publish_options pubopts = {}, v5::properties props = {});
```

The first one and the second one don't have life_keeper parameter. If we support the reference type of props then they also should have life_keeper parameter. What do you think?

NOTE: The third one have life_keeper. It is for topic_name and contents. At that time, I didn't care about props, but the life_keeper could be used for props.

Reply:
If we give an option that users can manage props lifetime by themselves, then the API should be

```cpp
    // approach A

    // life_keeper is optional
    void publish(packet_id_t packet_id, std::string topic_name, std::string contents, any life_keeper = any(), publish_options pubopts = {}, v5::properties props = {});
    void publish(packet_id_t packet_id, buffer topic_name, buffer contents, any life_keeper = any(), publish_options pubopts = {}, v5::properties props = {});
    // life_keeper is mandatory
    void publish(packet_id_t packet_id, as::const_buffer topic_name, as::const_buffer contents, any life_keeper, publish_options pubopts = {}, v5::properties props = {});

    // life_keeper is optional
    void pubrel(packet_id_t packet_id, v5::pubrel_reason_code reason_code = v5::pubrel_reason_code::success,  v5::properties props = {}, any life_keeper = any());
```

If we don't give it, in other words, users can use internal lifetime managed props only, then the API should be

```cpp
    // approach B

    // no life_keeper
    void publish(packet_id_t packet_id, std::string topic_name, std::string contents, publish_options pubopts = {}, v5::properties props = {});
    void publish(packet_id_t packet_id, buffer topic_name, buffer contents, publish_options pubopts = {}, v5::properties props = {});
    // life_keeper is mandatory, and it is for topic_name and contents
    void publish(packet_id_t packet_id, as::const_buffer topic_name, as::const_buffer contents, any life_keeper, publish_options pubopts = {}, v5::properties props = {});

    // no life_keeper
    void pubrel(packet_id_t packet_id, v5::pubrel_reason_code reason_code = v5::pubrel_reason_code::success,  v5::properties props = {});
```

Both approach A and B, no memory allocation happens in the sending process. In approach B, users cannot use the reference version of props. Users can user lifetime managed version of props only.

Reply:
I mean approach A allow all the following pattern, approach B allow the pattern1, 2, but 3 is not allowed. 

```
Pattern 1 (OK) Receive properties individual case. Or user create own property case.

+-----------------+   +-----------------+   +-----------------+  
|   prop1         |   |   prop2         |   |   prop3         |  
+-----------------+   +-----------------+   +-----------------+  
|    buffer-----+ |   |    buffer-----+ |   |    buffer-----+ |  
|     lifetime_ | |   |     lifetime_ | |   |     lifetime_ | |  
|      |        | |   |      |        | |   |      |        | |  
|      V        | |   |      V        | |   |      V        | |  
|     memory<---+ |   |     memory<---+ |   |     memory<---+ |  
+-----------------+   +-----------------+   +-----------------+  
                                                                 
                                                                 
                                                                 
Pattern 2 (OK) Receive properties all at once case
                                                                 
+-----------------+   +-----------------+   +-----------------+  
|   prop1         |   |   prop2         |   |   prop3         |  
+-----------------+   +-----------------+   +-----------------+  
|    buffer-----+ |   |    buffer-----+ |   |    buffer-----+ |  
|     lifetime_ | |   |     lifetime_ | |   |     lifetime_ | |  
+------|--------|-+   +------|--------|-+   +------|--------|-+  
       V        V            V        V            V        V    
+-------------------------------------------------------------+  
|                           memory                            |  
+-------------------------------------------------------------+  
                                                                 
                                                                 
Pattern 3 (NG) User uses buffer as pure view and actual memory is different place.

+-----------------+   +-----------------+   +-----------------+  
|   prop1         |   |   prop2         |   |   prop3         |  
+-----------------+   +-----------------+   +-----------------+  
|    buffer--------+  |    buffer--------+  |    buffer--------+ 
|     lifetime_   ||  |     lifetime_   ||  |     lifetime_   || 
|      |          ||  |      |          ||  |      |          || 
|      V          ||  |      V          ||  |      V          || 
|     nullptr     ||  |     nullptr     ||  |     nullptr     || 
+-----------------+|  +-----------------+|  +-----------------+| 
                   |                     |                     | 
                   |                     |                     | 
     Somewhere in  V  the user side      V                     V 
  +-------------------+ +-----------------+   +-----------------+
  |     memory        | |     memory      |   |     memory      |
  +-------------------+ +-----------------+   +-----------------+
```

Reply:
That's an excellent point.

I want to support users providing props with null shared_ptr, (externally managed lifetime).

So as a result, we should allow life_keeper parameter to all functions that may store the properties.

That includes publish, and pubrel.

I've updated the PR to support this with all publish and pubrel functions.

Please review again?

Reply:
> That's an excellent point.
> 
> I want to support users providing props with null shared_ptr, (externally managed lifetime).

Ok, that is Pattern 3 I wrote above. 

> So as a result, we should allow life_keeper parameter to all functions that may store the properties.
> That includes publish, and pubrel.

Ok, it is good.

> I've updated the PR to support this with all publish and pubrel functions.
> 
> Please review again?

I've reviewed.

I understand that you chose approach A in order to support Pattern 3.

```cpp
 // approach A

    // life_keeper is optional
    // OL1
    void publish(packet_id_t packet_id, std::string topic_name, std::string contents, any life_keeper = any(), publish_options pubopts = {}, v5::properties props = {});
    // OL2
    void publish(packet_id_t packet_id, buffer topic_name, buffer contents, any life_keeper = any(), publish_options pubopts = {}, v5::properties props = {});
    // life_keeper is mandatory
    // OL3
    void publish(packet_id_t packet_id, as::const_buffer topic_name, as::const_buffer contents, any life_keeper, publish_options pubopts = {}, v5::properties props = {});

    // life_keeper is optional
    void pubrel(packet_id_t packet_id, v5::pubrel_reason_code reason_code = v5::pubrel_reason_code::success,  v5::properties props = {}, any life_keeper = any());
```

But your update is a little bit different from my approach A. The difference is the order of parameters.

It's difficult to say witch is better. The point is life_keeper order. I add the tag OL1 to OL3 to discuss easily.

In my approach A, life_keeper is located on just after contents. It is consistent between OL1, OL2, and OL3. But for OL1 and OL2, life_keeper is for props, and props is located on **after** the life_keeper. It is unnatural for me.

* Pros
   * consistent
* Cons
   * Unnatural on OL1 and OL2

In your updated code, life_keeper is located on the last parameter in OL1 and OL2. In other words, just after props. But OL3 is not updated. So it is inconsistent. But I can say the following rule.

life_keeper is appeared just after the parameter that requires life_keeper.

All overloads satisfy the rule.

I think that there are three choices.

1. Use my approach A
2. Use your current updated and add the rule "life_keeper is appeared just after the parameter that requires life_keeper." in the document.
3. Modify OL3 that life_keeper to the last parameter.

I think that no 3 is the best. It keeps consistency with `async_*`. We can say life_keeper is always appear as the last parameter.

What do you think?

Reply:
> > That's an excellent point.
> > I want to support users providing props with null shared_ptr, (externally managed lifetime).
> 
> Ok, that is Pattern 3 I wrote above.

One comment. I understand that the Pattern 3 is important for you. Similarly Pattern 1 and Pattern 2 are very important for me. Please don't go a way to removing `buffer` from memory management mechanism in the future. Self memory managed `buffer` is key factor of my broker.

Reply:
> 3. Modify OL3 that life_keeper to the last parameter.

> I think that no 3 is the best. It keeps consistency with async_*. We can say life_keeper is always appear as the last parameter.

I think no. 3 is the best also.

I will update the PR.

Reply:
> One comment. I understand that the Pattern 3 is important for you. Similarly Pattern 1 and Pattern 2 are very important for me. Please don't go a way to removing buffer from memory management mechanism in the future. Self memory managed buffer is key factor of my broker.

Self managed buffer is also important to me, no plans to remove.

Unrelated: Can you give me your email, or email me at mike@meshplusplus.com ? I want to ask about your broker.

Reply:
redboltz@gmail.com My broker is propitiate. But I answer your question as long as I can do.

Reply:
For async_publish()

With the rule "Always last parameter", the function would look like

```cpp
    void async_publish(
        packet_id_t packet_id,
        as::const_buffer topic_name,
        as::const_buffer contents,
        publish_options pubopts = {},
        async_handler_t func = {},
        any life_keeper = {}
    ) {


    void async_publish(
        packet_id_t packet_id,
        as::const_buffer topic_name,
        as::const_buffer contents,
        publish_options pubopts,
        v5::properties props,
        async_handler_t func = {},
        any life_keeper = {}
    ) {

    void async_publish(
        packet_id_t packet_id,
        buffer topic_name,
        buffer contents,
        publish_options pubopts = {},
        async_handler_t func = {},
        any life_keeper = {}
    ) {

    void async_publish(
        packet_id_t packet_id,
        buffer topic_name,
        buffer contents,
        publish_options pubopts,
        v5::properties props,
        async_handler_t func = {},
        any life_keeper = {}
    ) {

    void async_publish(
        packet_id_t packet_id,
        std::string topic_name,
        std::string contents,
        publish_options pubopts = {},
        async_handler_t func = {},
        any life_keeper = {}
    ) {

    void async_publish(
        packet_id_t packet_id,
        std::string topic_name,
        std::string contents,
        publish_options pubopts,
        v5::properties props,
        async_handler_t func = {},
        any life_keeper = {}
    ) {
```

This means, that to call the function with async_handler, the user needs to do

```cpp
async_publish(
    5,
    "topic_1",
    "contents_1",
    publish_options{},
    v5::properties{},
    [](error_code ec) {
        /* my long callback function here */
    },
    any{});
```

With long callback function, it's hard to see the any{} at the end of the parameter list.

How about this rule?

> For sync functions, life_keeper is always last parameter.
> For async functions, life_keeper is always second last, before async_handler.

Reply:
> How about this rule?
> 
> > For sync functions, life_keeper is always last parameter.
> > For async functions, life_keeper is always second last, before async_handler.

I think that it is a good rule. It is also consistent to Boost.Asio's CompletionToken convention.
This requires more API updates. That means the updates increase a number of breaking changes, but I think that it is a good time to do this.

v7.0.1 is stable now so user can continue to use it, and the current master is already heading to v8.0.0.

I guess that after this updating, the APIs in the master would be stable.

Reply:
Yes, I hope, and believe, that API will remain stable after this.

Reply:
> For sync functions, life_keeper is always last parameter.
For async functions, life_keeper is always second last, before async_handler.

Just confirming the status. After you do the update above, I will merge the PR.

Reply:
Still working on it.

Unexpectedly busy.